File: 2_queue/单调队列Monoqueue/239. 滑动窗口最大值.ts
import { MaxDeque } from './MaxDequeue'

/**
 * `239. 滑动窗口最大值`
 */
function maxSlidingWindow(nums: number[], k: number): number[] {
  const queue = new MaxDeque<{ value: number }>()
  const res = []

  for (let i = 0; i < nums.length; i++) {
    queue.append({ value: nums[i] })
    if (i >= k) queue.popLeft()
    if (i >= k - 1) res.push(queue.max)
  }

  return res
}

export {}

File: 11_动态规划/dp优化/辅助数据结构dp/单调队列优化dp/单调队列Monoqueue/239. 滑动窗口最大值.py
from typing import List
from MaxQueue import MaxQueue


def maxSlidingWindow(nums: List[int], k: int) -> List[int]:
    """滑动窗口最大值"""
    queue = MaxQueue()
    res = []
    n = len(nums)
    for right in range(n):
        queue.append(nums[right])
        if right >= k:
            queue.popleft()
        if right >= k - 1:
            res.append(queue.max)
    return res


if __name__ == "__main__":
    print(maxSlidingWindow([1, 3, -1, -3, 5, 3, 6, 7], 3))
